style.md

```diff
@@ -1557,7 +1557,237 @@ type Client struct {
 </td></tr>
 </tbody></table>
 
-### Use Field Names to initialize Structs
+As a general rule of thumb, embedding should not make types more difficult to
+use, and should be viewed as an "is-a" declaration: the embedding (outer) type
+can function as the embedded (inner) type, and therefore "is a" form of the
+inner type. A type embedding a `sync.Mutex` declares to users that it "is a"
+mutex; a type embedding a `bytes.Buffer` declares that it "is a" buffer.
+
+Embedded types' methods or fields should never be shadowed by first-class
+methods or fields defined as part of the embedding type, and embedding should
+never be used for callsite brevity or other cosmetic convenience.
+
+When embedding, visibility and type semantics should be taken into account:
+because exported fields and methods of an embedded type will also be exported
+by the embedding type, careless use can bleed implementation details through
+types' API boundaries, expose internal type faculties that can be used in an
+unsafe or unknown way, or change the semantics of the embedding type altogether.
+
+All of the guidance in this section applies primarily to embedding types that
```

Comment:
Should we say an exported type should not embed an unexported type?

Reply:
The unfortunate reality is probably "it depends":

- Is the zero type already useful? If yes, and the unexported type does not have a useful (and in this case, zero-touch) zero value, then that's bad.
- If the type now requires complex/opaque construction as a result of embedding the unexported type, then that's bad.
- If the embedded type's own fields are exported, that's bad.

I think if we established a basic guideline of "An embedded type must not impede existing trivial use of the embedding type" (i.e. to go from trivially constructable/copyable to not), that might be enough to discourage misuse while allowing for design choices that we don't have the ability to foresee.

This is all with the understanding that, in most cases, simply not embedding the type is the right answer. So maybe that's the right guidance:

> By default, embedding unexported types should be avoided. Embedded types must also not make the embedding type's zero value less useful, or make its construction opaque to the user.

Thoughts?